docs/src/references/mqtt-api.md

```diff
@@ -0,0 +1,838 @@
+|&lt;cmd_id&gt;|Unique command id which is unique for the command instance. e.g. `123456`, `d511a86cab95be81` etc.|
+
+### Command examples
+
+The following table details some example command types which are supported by thin-edge.
+
+|Command Type|Example Topic|
+|------|---|
+|software_list|`te/<identifier>/cmd/software_list/<cmd_id>`|
+|software_update|`te/<identifier>/cmd/software_update/<cmd_id>`|
+|config_snapshot|`te/<identifier>/cmd/config_snapshot/<cmd_id>`|
+|config_update|`te/<identifier>/cmd/config_update/<cmd_id>`|
+|firmware_update|`te/<identifier>/cmd/firmware_update/<cmd_id>`|
+|restart|`te/<identifier>/cmd/restart/<cmd_id>`|
+
+The command would be interpreted differently based on the target entity.
+For example, the `restart` could mean either a device restart or a service restart based on the target entity.
+
+### Examples: With default device/service topic semantics
+
+#### Command to main device
+
+Command to fetch the software list from the main device:
+
+```sh te2mqtt
+tedge mqtt pub -r 'te/device/main///cmd/software_list/123' '{
+  "status": "init"
+}'
+```
+
+The `status` value will transition from `init` to the final `successful` or `failed` via many intermediate states
+such as `validating`, `downloading`, `executing` etc.
+
+#### Command to child device
+
+Command to update the firmware of a child device:
+
+```sh te2mqtt
+tedge mqtt pub -r 'te/device/child01///cmd/firmware_update/123' '{
+  "status": "init",
+  "attempt": 1,
+  "name": "OpenWRT",
+  "version": "22.03",
+  "url":"http://127.0.0.1:8000/tedge/file-transfer/tedge-child/firmware_update/93d50a297a8c235",
+  "sha256":"c036cbb7553a909f8b8877d4461924307f27ecb66cff928eeeafd569c3887e29"
+}'
+```
+
+#### Command to a service
+
+Command to update the configuration of a service:
+
+```sh te2mqtt
+tedge mqtt pub -r 'te/device/main/service/collectd/cmd/config_update/123' '{
+  "status": "init",
+  "type": "collectd"
+  "path": "/etc/collectd/collectd.conf",
+  "url": "http://127.0.0.1:8000/tedge/file-transfer/collectd/config_update/collectd",
+}'
+```
+
+### Commands metadata
+
+For commands as well, additional command specific metadata can be registered as retained messages on the `<cmd_type>` topic.
+
+For example, the supported configuration list of the main device can be declared as follows:
+
+```sh te2mqtt
+tedge mqtt pub -r te/device/main///cmd/config_snapshot '{
+  "description": "Upload a configuration from the device",
+  "supportedTypes": ["mosquitto", "tedge", "collectd"]
+}'
+```
+
+*TODO*: Unified config list for both `config_snapshot` and `config_update` commands or support different lists for both?
+
+## Health check
+
+Services can publish their health status as follows:
+
+```sh te2mqtt
+tedge mqtt pub -r te/device/main/service/tedge-agent/cmd/health/ '{
+  "pid": 1234,
+  "status": "successful",
+  "time": 1674739912
+}'
```

Comment:
The exact details of the health check endpoint are still being finalized in another PR, https://github.com/thin-edge/thin-edge.io/pull/2071

So this area is subject to change. I'm ok if we leave it how it is for now, and we can updated the command aspects after https://github.com/thin-edge/thin-edge.io/pull/2071 is finalized.